cores/esp8266/FunctionalInterrupt.cpp

```diff
@@ -8,7 +8,7 @@ typedef void (*voidFuncPtr)(void);
 typedef void (*voidFuncPtrArg)(void*);
 
 // Helper functions for Functional interrupt routines
-extern "C" void ICACHE_RAM_ATTR __attachInterruptArg(uint8_t pin, voidFuncPtr userFunc, void*fp , int mode);
```

Comment:
After reading several times, I don't understand why it is needed to have this `bool functional`.
It is always checked along with the not-`nullptr` arg. Why is `arg` not sufficient ?

Reply:
I will do a detailed check (maybe thursday. otherwise friday).
I needed it in ESP32 to distinguish the use of attachinterruptArg from user exposed and usage for functional.

Reply:
@d-a-v The `bool functional` prevents undefined type-casting` void* arg` to `ArgStructure*` in all those cases that the whole exercise of exposing
`void attachInterruptArg(uint8_t pin, void (*)(void*), void * arg, int mode)`
is done for: whenever it's not an `ArgStructure*`, not null either, but for instance some `this` pointer or whatever else fits into `void*` (32bit integers).
I've revisted the code and I am quite sure it's needed where it is.

Reply:
FYI: I've tested using microsecond resolution instead of CPU cycles for the Software Serial bit timing. The ESP8266 shows no difference, but on the ESP32, at high bit rate (57600cps), the error rate increases measurably.

Reply:
Given the reduction of the complexity of the generated code, I always try to cpu cycle (if measured durations are less than 26 seconds (esp8266@160MHz) or if overflow is managed)

Reply:
But here (SW serial) it impacts basic usefulness, each worsening in error rate renders it useless in more cases.
My point of view is that unconditionally requiring C++ functional objects in the attachInterruptArg interface through an API that looks like plain C attracts confusion. The compiler can't catch any of it. I have to admit, that I am not entirely happy with the whole construct between FunctionalInterrupt and core_esp8266_wiring_digital, exemplified by the need to copy&paste type definitions
`//duplicate from functionalInterrupt.h keep in sync
typedef struct InterruptInfo`
because the relationship is bit forced as it is right now.
But anyway, in ESP32 Arduino, it's just the same, so future patches just might take advantage or a more in similarity.